File: Leetcode July Challenge/C++/Binary Tree Pruning.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int helper(TreeNode* root){
        if(root == NULL){
            return 0;
        }
        
        if(root-> left == NULL && root->right == NULL){
            if(root->val == 0){
                return 0;
            }else{
                return 1;
            }
        }
        
        int leftSubTreeOnes = helper(root->left);
        int rightSubTreeOnes = helper(root->right);
        
        if(leftSubTreeOnes == 0){
            root->left = NULL;
        }
        if(rightSubTreeOnes == 0){
            root->right = NULL;
        }
        
        return leftSubTreeOnes + rightSubTreeOnes + (root->val == 1 ? 1:0);
        
    }
    
    TreeNode* pruneTree(TreeNode* root) {
        
        if(helper(root) == 0){
            return NULL;
        }
        
        return root;
        
    }
};

File: Leetcode July Challenge/Python/Binary Tree Pruning.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def helper(self, root):
        if root == None:
            return 0
        
        if root.left == None and root.right == None:
            if root.val == 0:
                return 0
            else:
                return 1
        
        leftSubTreeOnes = self.helper(root.left)
        rightSubTreeOnes = self.helper(root.right)
        
        if leftSubTreeOnes == 0:
            root.left = None
            
        if rightSubTreeOnes == 0:
            root.right = None
        
        if root.val == 1:
            
            return leftSubTreeOnes + rightSubTreeOnes + 1
        
        else:
            return leftSubTreeOnes + rightSubTreeOnes
        
    
    def pruneTree(self, root: TreeNode) -> TreeNode:
        if self.helper(root) == 0:
            return None
        else:
            return root
        
